blosc/shuffle.c

```diff
@@ -7,585 +7,278 @@
-/* Unshuffle a block.  This can never fail. */
-void _unshuffle(size_t bytesoftype, size_t blocksize,
-                       const uint8_t* _src, uint8_t* _dest)
-{
-  size_t i, j, neblock, leftover;
-
-  /* Non-optimized unshuffle */
-  neblock = blocksize / bytesoftype;  /* Number of elements in a block */
-  for (i = 0; i < neblock; i++) {
-    for (j = 0; j < bytesoftype; j++) {
-      _dest[i*bytesoftype+j] = _src[j*neblock+i];
-    }
-  }
-  leftover = blocksize % bytesoftype;
-  memcpy(_dest+neblock*bytesoftype, _src+neblock*bytesoftype, leftover);
-}
 
+/*  Define function pointer types for shuffle/unshuffle routines. */
+typedef void(*shuffle_func)(size_t, size_t, const uint8_t* const, uint8_t* const);
+typedef void(*unshuffle_func)(size_t, size_t, const uint8_t* const, uint8_t* const);
 
-#if defined(__SSE2__)
-/* #pragma message "Using SSE2 version shuffle/unshuffle" */
+/* An implementation of shuffle/unshuffle routines. */
+typedef struct shuffle_implementation {
+  /* Name of this implementation. */
+  const char* name;
+  /* Function pointer to the shuffle routine for this implementation. */
+  shuffle_func shuffle;
+  /* Function pointer to the unshuffle routine for this implementation. */
+  unshuffle_func unshuffle;
+} shuffle_implementation_t;
 
-/* The SSE2 versions of shuffle and unshuffle */
 
-#include <emmintrin.h>
+/*  Detect hardware and set function pointers to the best shuffle/unshuffle
+    implementations supported by the host processor. */
+#if defined(SHUFFLE_AVX2_ENABLED) || defined(SHUFFLE_SSE2_ENABLED)    /* Intel/i686 */
 
-/* The next is useful for debugging purposes */
-#if 0
-static void printxmm(__m128i xmm0)
-{
-  uint8_t buf[16];
+#if defined(_MSC_VER) && !defined(__clang__)
+  #include <immintrin.h>  /* Needed for _xgetbv */
+  #include <intrin.h>     /* Needed for __cpuid */
+#else
 
-  ((__m128i *)buf)[0] = xmm0;
-  printf("%x,%x,%x,%x,%x,%x,%x,%x,%x,%x,%x,%x,%x,%x,%x,%x\n",
-          buf[0], buf[1], buf[2], buf[3],
-          buf[4], buf[5], buf[6], buf[7],
-          buf[8], buf[9], buf[10], buf[11],
-          buf[12], buf[13], buf[14], buf[15]);
+/*  Implement the __cpuid and __cpuidex intrinsics for GCC, Clang,
+    and others using inline assembly. */
+__attribute__((always_inline))
+static inline void
+__cpuidex(int32_t cpuInfo[4], int32_t function_id, int32_t subfunction_id) {
+  __asm__ __volatile__ (
+# if defined(__i386__) && defined (__PIC__)
+  /*  Can't clobber ebx with PIC running under 32-bit, so it needs to be manually restored.
+      https://software.intel.com/en-us/articles/how-to-detect-new-instruction-support-in-the-4th-generation-intel-core-processor-family
+  */
+    "movl %%ebx, %%edi\n\t"
+    "cpuid\n\t"
+    "xchgl %%ebx, %%edi":
+    "=D" (cpuInfo[1]),
+#else
+    "cpuid":
+    "=b" (cpuInfo[1]),
+#endif  /* defined(__i386) && defined(__PIC__) */
+    "=a" (cpuInfo[0]),
+    "=c" (cpuInfo[2]),
+    "=d" (cpuInfo[3]) :
+    "a" (function_id), "c" (subfunction_id)
+    );
```

Comment:
for gcc you can use the intrinsic that was used before this PR, though there is probably no way around asm on windows. For linux in general you probably somethimes have `cpuid.h` but I have never tested that with clang.